lib/AST/ASTImporter.cpp

```diff
@@ -3001,32 +3001,46 @@ ExpectedDecl ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
       }
 
       if (FunctionDecl *FoundFunction = dyn_cast<FunctionDecl>(Found)) {
-        if (FoundFunction->hasExternalFormalLinkage() &&
-            D->hasExternalFormalLinkage()) {
-          if (IsStructuralMatch(D, FoundFunction)) {
-            const FunctionDecl *Definition = nullptr;
-            if (D->doesThisDeclarationHaveABody() &&
-                FoundFunction->hasBody(Definition))
-              return Importer.MapImported(
-                  D, const_cast<FunctionDecl *>(Definition));
-            FoundByLookup = FoundFunction;
-            break;
-          }
 
-          // FIXME: Check for overloading more carefully, e.g., by boosting
-          // Sema::IsOverload out to the AST library.
+        bool VisibilityMatch = true;
+        if (D->hasExternalFormalLinkage())
+          VisibilityMatch = FoundFunction->hasExternalFormalLinkage();
+        else {
```

Comment:
I think you could simplify this by uplifting the TU match and by starting with the anonymous namespace case first, something like:
```
        else if (Importer.GetFromTU(FoundFunction) == D->getTranslationUnitDecl()) {
          if (D->isInAnonymousNamespace())
            VisibilityMatch = FoundFunction->isInAnonymousNamespace()                  
          else
             VisibilityMatch = !FoundFunction->isInAnonymousNamespace() &&
                              !FoundFunction->hasExternalFormalLinkage()            
                           

```